Add SCA scripts to KPDB pipeline + fix borough geometry assignment

mbh329 opened the pull request:
## motivations

- resolves issue #399.
- The original motivation behind this PR was to get the existing SCA aggregate scripts (that were written a few years ago and executed in carto)  running so that we can deliver outputs to Housing/Capital Planning and they would perform any necessary clean up and formatting work before it was delivered to SCA.
- To Incorporate them into the KPDB build process I created a separate `04_sca_aggregate.sh` to handle the execution and creation of the scripts and subsequent output files (`longform_csd_output` , `longform_es_zone_output`, `longform_subdist_output_cp_assumptions`). As of now this is NOT incorporated into the build workflow/Github Action and can only be done locally.

## changes
- These 3 scripts are essentially the same scripts that we received with some minor tweaks to the table names so that they are pulling in KPDB specific tables which are "like for like" input tables (e.g. `kpdb` -> `_kpdb`, `Doe_school_zones_es_2019` -> `doe_eszones`)
- Add in a separate script that preprocesses the input source data needed to run the scripts, creates a table called `zap_project_many_bbls` that was needed and appeared in each script, runs the each of the 3 scripts (`boundaries_es_zone.sql`, `boundaries_school_districts.sql`, and `boundaries_school_subdistricts.sql`) and then exports the CSV's. I decided to keep this separate from the existing `build` and `export` process because this is typically the very last thing that DE or Housing does. I don't think it makes sense to run the scripts and write new outputs each time we rebuild KPDB locally. 
- Update the `dev` branch on the main KPDB repo with the latest `data` submodule to avoid confusion. This has been signed off by housing as the "final' version of KPDB with all the latest corrections.
- I updated the logic to apply the borough to a specific record based on that records geometry by taking the `dcp_boroboundary` file without `water included` and used `st_intersect` instead of `st_within`. The motivation behind this was "The issue was that when I was using `st_intersect` with the water included boundaries, Manhattan's water boundary is right up to the land boundary of Brooklyn so all the records on the BK waterfront with overlap into Manhattan were getting assigned to Manhattan".
- 
## notes
- This should be rerun locally by another engineer to make sure it works on their machine. 
- The sql scripts related to the SCA aggregate tables are a mess (weird formatting, etc) but I was hesitant to change the formatting due to the tight timeline. Once we have turned the page on this chapter of KPDB, we should go and reformat those scripts in another PR. Issue created here #401
- 

## todo

- [ ] ...

mbh329 committed:
import sca geography tables into dataloading

mbh329 committed:
add files to pipeline to run sca aggregation

mbh329 committed:
update scripts and add export step in sca aggregate

mbh329 committed:
test ST_intersects to assing borough

mbh329 committed:
update get boro and dataloading with new logic

fvankrieken committed:
pull in latest data

mbh329 committed:
remake zap_project_many_bbls table

mbh329 committed:
add sca aggregate tables

fvankrieken commented on sca_aggregate/create_zap_projects.sql:
```diff
@@ -0,0 +1,35 @@
+--- Get dcp_bblnumber and dcp_name from dcp_projectbbls
+DROP TABLE IF EXISTS dcp_projectbbl_tmp;
+CREATE TABLE dcp_projectbbl_tmp AS
```
Doesn't need to change, just more of an fyi - if these are truly temp tables and you want them treated as such, you can do `CREATE TEMPORARY TABLE my_table AS`. The table is automatically dropped at the end of the current session. Downside obviously is then you can't query it separately to investigate if something is awry in your final result

mbh329 replied:
Ahh yeah. I originally thought this was going to be a tmp table but then thought it might be good to have passed the current session, I can just change the name of table

fvankrieken commented on sql/_functions.sql:
```diff
@@ -23,6 +23,6 @@ $$ LANGUAGE sql;
-    FROM dcp_boroboundaries_wi b
-    WHERE ST_Within(_geom, b.wkb_geometry)
+    FROM dcp_boroboundaries b
+    WHERE ST_Intersects(_geom, b.wkb_geometry)
```
Do we have cases in this dataset where geoms might intersect more than one boro? Do you know what will happen in that case with this function - will it return a single varchar as the type signature suggests (first boro it finds?), or will it throw an error?

Also, a bit unsure if this really should be a function - it's basically just left joining dcp_boroboundaries to the geom column that this function is being applied to, I would imagine just doing that join would be more performant (as long as there's an index on the geom column). It's also only called in `create_kpdb`, so I would lean with that personally

fvankrieken replied:
Actually didn't know the answer to my first question, so did some testing locally - 
``` sql
drop if exists table "test";
create temp table "test" ("a", "b") as values (1, 2), (1, 1);
select * from test;
create or replace function test_function(val int)
returns int as $$
    select b from test where a=val
    $$ language sql;

select test_function(a), b from "test";
```

returns 
| test_function | b |
| -- | -- |
| 2 | 2 |
| 2 | 1 |

So seems to just be grabbing one of the values where it matches, in this case the "last" though I don't know how much to read into that.

fvankrieken replied:
Not saying second is ideal, but at least then we can actually tell when we're intersecting multiple boros. The first we're potentially somewhat randomly (or at least indiscriminately) losing information in these cases

mbh329 replied:
I don't think so. The project polygons look to be exclusively in the regular borough boundaries but this was an issue when we were joining on the `dcp_boroboundaries_wi` water included boundarie of Manhattan borders the land border of brooklyn so all of the piers that are being redeveloped on the BK waterfront were being misclassified as Manhattan records because they intersected both BK and Manhattan. Previously the KPDB pipeline was bringing in the water included file and then using st_within which was creating issues because the projects weren't always totally within the borough boundary file. 

Yeah I think from a performance side its probably best to drop the function and just do a join but I think maybe we can do that when we refactor KPDB, I'm overly cautious changing anything at this point and just want it to be done. I also think that we might need to make some adjustments to how we assign boroughs for records without a geometry which hopefully will be ironed out when we meet at 2pm today

fvankrieken replied:
So tried running both ways

Current method, ~9 min to query on my machine
27989	4
23815	3
18306	5
9731	2
4552	1

Add index, left join dcp_boroboundaries, ~4 min to query on my machine
27989	4
23823	3
18306	5
9731	2
4552	1

So looks like we have 8 records on the brooklyn queens border that intersect both, all of which are getting assigned to queens. I'll check these out after lunch

fvankrieken replied:
Following records
| source | record_id | record_name |
|-|-|-|
|DCP Application | 2021K0237 | River Ring|
|DCP Application | 2021K0313 | 1 Java Street Waterfront|
|DCP Application | 2023K0138 | Herkimer - Williams - City Map Change|
|DCP Application | P2014K0254 | Williamsburg Edge Phase 2 Waterfront Certification|
|DOB | 320377463 | 130 FURMAN STREET|
|DOB | 320513182 | 90 FURMAN STREET|
|DOB | B00539478 | 18 INDIA STREET|
|DOB | B00629573 | 105 RIVER STREET|

fvankrieken commented:
A bit of cleanup - I would just delete the old scq sql files (5a, 5c, etc) and maybe update the comments (or mostly delete) at the top of our sca aggregate files

fvankrieken commented:
Seems to all be building cleanly for me locally 👍

mbh329 committed:
rename data table

mbh329 committed:
clean up comments at top of scripts

mbh329 committed:
remove old scripts

mbh329 committed:
add school subdistzone to final output

fvankrieken commented:
Copying from #403 

After fixes in said PR, 9 records are still having no boro assigned

| source | record_id | record_name |
|-|-|-|
| HPD Projected Closings | 3d3a2dd3c6ec864f5e18cec8e03eb2a7 | 784 COURTLANDT AVENUE |  
| HPD Projected Closings | ff4c3b3e09d4c3a0d527a5c7988242d0 | 1559 PROSPECT PLACE |  
| HPD RFPs | 0ce4a0321f576ad9b397dec9358f255f | MWBE Site B - 1921 Atlantic Ave |  
| HPD RFPs | 100c52f92560bc161659f99b08340e9f | 97 West 169 St |  
| HPD RFPs | 1db4d18c803723face138966e70095d8 | NYCHA West Brighton RFP |  
| HPD RFPs | 473c582483789bb07b5837ecdb746ae4 | Bedford-Stuyvesant Community Wealth and Wellness RFP (Fulton-Saratoga) |  
| HPD RFPs | 862c3d3640912f5ec6561375a03449ef | Hunters Point F + G |  
| HPD RFPs | 9bd199e88bb8dd9b128d0d3c71b520b5 | Hunters Point F + G |  
| HPD RFPs | defae821da6ac0b1a7610be5c7d5f484 | SustaiNYC (E. 111th Street) |  

We'll likely need to add these to the corrections

fvankrieken committed:
special cases to parse borough from record_id (#403)

- fix slight change in logic I introduced to sca school district script, then also get districts at end of script for still unassigned projects
- in kpdb build, when no boro assigned, get from project_id when possible

fvankrieken commented:
Merged my changes into this branch @mbh329. I'm going to rebuild from scratch locally now on this branch

Linking @AmandaDoyle's comment on my PR [here](https://github.com/NYCPlanning/db-knownprojects/pull/403#issuecomment-1489342948) - waiting on confirmation but then we'll need to apply corrections

fvankrieken committed:
error arrising from vacuum in transaction block

mbh329 commented:
rebuilding locally now @fvankrieken

mbh329 commented:
ran into a storage issue, cleaning stuff up and will try to rerun

fvankrieken committed:
add subdistrict to second aggregate query

fvankrieken committed:
add sca option to kpdb.sh

damonmcc commented on bash/01_dataloading.sh:
```diff
@@ -16,9 +16,14 @@ import_private dcp_dcpprojectteams &
 # Load other tables
 import_public dcp_mappluto_wi &
-import_public dcp_boroboundaries_wi &
+import_public dcp_boroboundaries & #Don't pull the _wi boundaries 
```
no need for the comment?

damonmcc commented on sql/combine.sql:
```diff
@@ -366,5 +366,6 @@ FROM(
\ No newline at end of file
+);
+
+CREATE INDEX combined_geom_idx ON combined USING GIST(geom);
```
just curious, is this to speed up subsequent uses of of the table `combined`?

fvankrieken replied:
Yup, there's a fairly long-running join later on

damonmcc reviewed (approved):
LGTM especially considering the urgency!

is it worth running locally before merging to ensure the csvs in `sca_output/` are definitely the result of the latest code changes? looks like they were generate 2 days ago

fvankrieken commented:
> LGTM especially considering the urgency!
> 
> is it worth running locally before merging to ensure the csvs in `sca_output/` are definitely the result of the latest code changes? looks like they were generate 2 days ago

Yes, we should commit latest generated sca outputs. Or @mbh329 are those supposed to go in the data repo instead?

mbh329 commented:
I've been generating them separately from the data repo, I think we can commit the most recent sca agg tables and then merge. Would you like to do that or should I? I have the latest ones from this morning but can rerun to make sure

fvankrieken commented:
> I've been generating them separately from the data repo, I think we can commit the most recent sca agg tables and then merge. Would you like to do that or should I? I have the latest ones from this morning but can rerun to make sure

I just dropped my local kpdb db to make room for edde so please do!

mbh329 committed:
remove comment

mbh329 committed:
Merge branch '399-add-scripts-to-kpdb-pipeline' of github.com:NYCPlanning/db-knownprojects into 399-add-scripts-to-kpdb-pipeline

mbh329 commented:
rebuilding now with latest data/corrections

mbh329 commented:
Taking longer than I would like, ran into some more obtuse memory issues but building now

mbh329 commented:
![image](https://user-images.githubusercontent.com/41444592/228924465-d1d8a84f-e083-4163-828d-0c10bab18c45.png)

fvankrieken commented:
I didn't have any issue earlier - do you want me to build locally?

mbh329 commented:
Still running the aggregate step but the corrections are not being applied to the 9 records without a borough

mbh329 commented:
@AmandaDoyle @fvankrieken @damonmcc

mbh329 committed:
add latest csv files of sca agg and outputs, add latest data

fvankrieken committed:
update build to apply borough corrections, regenerate sca tables

fvankrieken committed:
update output

fvankrieken committed:
399 fvk edits (#404)

- changes to export script
- removal of export in sca agg script
- renaming of columns in sca agg so that they are consistent and _kpdb table isn't altered before export

fvankrieken committed:
update data ref
